scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -807,26 +843,38 @@ def run_and_propagate(self):
         """
         assert self.__k8s_cache is not None
 
+        consecutive_warm_pods = 0
         while self._run_state.is_running():
-            container_id, pod_namespace, pod_name = self.__pick_next_pod()
-            if container_id is not None:
-                if self.is_warm(pod_namespace, pod_name):
+            try:
+                container_id, pod_namespace, pod_name = self.__pick_next_pod()
+                if container_id is not None and not self.is_warm(pod_namespace, pod_name):
+                    consecutive_warm_pods = 0
+                    try:
+                        options = ApiQueryOptions(max_retries=self.__max_query_retries, rate_limiter=self.__rate_limiter)
+                        self.__k8s_cache.pod(pod_namespace, pod_name, allow_expired=False, query_options=options)
+                        self.__record_warming_result(container_id, success=True)
+                    except K8sApiPermanentError, e:
+                        self.__record_warming_result(container_id, permanent_error=e,
+                                                     traceback_report=traceback.format_exc())
+                    except K8sApiTemporaryError, e:
+                        self.__record_warming_result(container_id, temporary_error=e,
+                                                     traceback_report=traceback.format_exc())
+                    except Exception, e:
+                        self.__record_warming_result(container_id, unknown_error=e,
+                                                     traceback_report=traceback.format_exc())
+                else:
                     # Something else warmed the pod before we got a chance.  Just take the win.
-                    continue
+                    self.__record_warming_result(container_id, already_warm=True)
+                    consecutive_warm_pods += 1
+                    global_log.log(scalyr_logging.DEBUG_LEVEL_3,
+                                   'CacheWarmer.run_and_propagate: %s consecutive warms in a row: %s %s %s'
+                                   % (consecutive_warm_pods, container_id, pod_namespace, pod_name))
+                    if consecutive_warm_pods == 10:
+                        # TODO: Do not remove this circuit breaker until this issue is fully understood.
```

Comment:
There are some outstanding TODOs that must be resolved before merging back to master. This is one of them.